### PR TITLE
Panels: fix loading panels with non-array targets

### LINKED ISSUE #15993: Non-metric panel fails dashboard to load in master
I'm trying to load a dashboard containing Problems panel from Zabbix plugin. Dashboard fails to load with error:
Screenshot from 2019-03-14 12-29-13
```
TypeError: Cannot read property 'refId' of undefined
    at PanelModel../public/app/features/dashboard/state/PanelModel.ts.PanelModel.ensureQueryIds (PanelModel.ts:130)
```
Grafana built from `master` at 4bbd6ab79cc3b023a8e671eb8fe21f007a492639

Panel has option `"dataFormats": [],` in the `plugin.json` file, so in this case `PanelModel` shouldn't validate queries:
https://github.com/grafana/grafana/blob/9e7d1f42753495254c7a9220161b0818decbf58a/public/app/features/dashboard/state/PanelModel.ts#L122

### PR DESCRIPTION
**Pull request by alexanderzobnin**
This PR fixes #15993
Bug happens when panel has not array-like `targets` object (this may be true for non-metrics panel). Root cause - `ensureQueryIds()` function in `PanelModel` module iterates over targets with `for ... of` loop which can handle any iterable types (objects as well).
https://github.com/grafana/grafana/blob/9e7d1f42753495254c7a9220161b0818decbf58a/public/app/features/dashboard/state/PanelModel.ts#L129
I suggest to use normal `for ...` loop and iterate only over array elements.

### CODE REVIEW

File: public/app/features/dashboard/state/PanelModel.ts [hunk 1]
```diff
@@ -126,7 +126,8 @@ export class PanelModel {
 
   ensureQueryIds() {
     if (this.targets) {
-      for (const query of this.targets) {
+      for (let i = 0; i < this.targets.length; i++) {
+        const query = this.targets[i];
```
Comment (torkelo) [inline public/app/features/dashboard/state/PanelModel.ts]: this fix feels fragile, especially without a test. Maybe check if it's an array instead and keep the other for of loop. Not sure why this makes a difference, for of loops generate a for loop like above so this fix should not make a difference.

Comment (alexanderzobnin) [reply public/app/features/dashboard/state/PanelModel.ts]: The problem is that in my case targets contains some artifacts - it's array-like object with additional props:

Screenshot from 2019-03-15 10-22-29

`for ... of` returns first element of array (I think it treats my object as array with length 3) and then returns `undefined` on 2nd and 3rd iterations.

Comment (alexanderzobnin) [reply public/app/features/dashboard/state/PanelModel.ts]: I tried to use 
```js
      for (const query of this.targets) {
        if (query && !query.refId) {
          query.refId = this.getNextQueryLetter();
        }
      }
```
but it causes infinite loop for my object.

Comment (alexanderzobnin) [reply public/app/features/dashboard/state/PanelModel.ts]: This approach works as well:
```ts
  ensureQueryIds() {
    if (this.targets && _.isArray(this.targets)) {
      for (const query of this.targets) {
        if (!query.refId) {
          query.refId = this.getNextQueryLetter();
        }
      }
    }
  }
```

**Comment by alexanderzobnin** [comment]
I think it's webpack bug actually. If `this.targets` is a plain object it shouldn't be iterable with `for of`. Webpack produces this output after compilation:
```js
PanelModel.prototype.ensureQueryIds = function () {
    var e_1, _a;
    if (this.targets) {
        try {
            for (var _b = tslib__WEBPACK_IMPORTED_MODULE_0__["__values"](this.targets), _c = _b.next(); !_c.done; _c = _b.next()) {
                var query = _c.value;
                console.log(query);
                if (!query.refId) {
                    query.refId = this.getNextQueryLetter();
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
};
```
`__values()` function is:
```js
export function __values(o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
}
```
Look at this string: `return { value: o && o[i++], done: !o };`
It just returns `o[0]`, then `o[1]` etc. Since `o.length` is `undefined` for objects, it produces infinite loop.

**Comment by alexanderzobnin** [comment]
I opened related issue in TypeScript repo:
https://github.com/Microsoft/TypeScript/issues/30423

**Comment by alexanderzobnin** [comment]
Cool!

**Pull request merged**